### ISSUE TITLE
int8 mode only 5-10% faster than fp16

### ISSUE THREAD

**Issue by mstrfx** [NVIDIA/TensorRT#585]
## Description

So I did int8 calibration on yolov3 onnx model and was expecting at least 30% speed improvement. However, inference time difference is negligible. 
![compilation](https://user-images.githubusercontent.com/11006070/83369564-6b29ae80-a3c5-11ea-899d-5526a019f017.png)
So is this normal?

## Environment

**TensorRT Version**: 7
**GPU Type**: rtx 2080 ti / tesla t4
**Nvidia Driver Version**: 
**CUDA Version**: 10.2
**CUDNN Version**: 7.6.5
**Operating System + Version**: Ubuntu 18.04
**Python Version (if applicable)**: 3.6.9
**TensorFlow Version (if applicable)**: 
**PyTorch Version (if applicable)**: 1.5
**Baremetal or Container (if container which image + tag)**: 

## Relevant Files

https://gist.github.com/mstrfx/b4694c74b43ada91a76b665e8b0d34f7

## Steps To Reproduce

convert yolov3 model to onnx
```
python3 calib_onnx.py int8 (or fp16)
python3 benchmark.py int8 <batch_size>
```

**Comment by le8888e** [comment]
Hi @mstrfx,
    I came up with the same problem with you. My model is an onnx model for text detection and I used C++ API, INT8 runs almost the same speed as FP16.
    Furthermore, in my case INT8 and FP16 runs only 10% faster than FP32, which is much slower than I expected. Do you measure the speed difference between IN8 and FP32? 
    Thank you.

**Comment by mstrfx** [comment]
hey @le8888e  
Depends on batch size but I get inference time around 30-50% lower. I was just expecting int8 to be at least 20% faster than fp16

**Comment by rmccorm4** [comment]
Hi @mstrfx ,

I believe if you build your engine with the logger set to VERBOSE, you can see when int8 kernels are chosen over fp32/fp16 kernels (try saving/redirecting the logger output to a file and searching for "int8"). It's possible that only a small percentage of the network was eligible/faster with int8 - in this case the overall speedup wouldn't be too great.

You can probably also use a profiling tool such as nsight (newer) or nvprof (older) to see where the bottlenecks are. I believe these tools will also tell you more details about how many int8 kernels were used vs. fp16 vs. fp32, etc.

**Comment by le8888e** [comment]
Hi @rmccorm4 ,

In my case the overall speed up of INT8 is negligible.
I try to build my engine with the logger set to VERBOSE and here is my logger when build INT8 engine :
[int8.txt](https://github.com/NVIDIA/TensorRT/files/4733540/int8.txt)
Does it indicates that conv layer runs in FP32 mode?
Thank you.

**Comment by jkjung-avt** [comment]
I shared my results applying INT8 TensorRT optimization on yolov3/yolov4 models in my [jkjung-avt/tensorrt_demos](https://github.com/jkjung-avt/tensorrt_demos#int8_and_dla) repository.

For example, I tested TensorRT YOLOv3 engines on Jetson Xavier NX (JetPack-4.4: CUDA 10.2, cuDNN 8 and TensorRT 7.1.3) and got the following frames-per-second (FPS) numbers.  (Note these FPS numbers include all of image preprocessing, TensorRT inference, postprocessing and display.)

   | TensorRT engine  |   FP16   |   INT8   |   DLA0   |   DLA1   |
   |:-----------------|:--------:|:--------:|:--------:|:--------:|
   | yolov3-608       |   15.2   |   23.1   |   14.9   |   14.9   |

In this case, the **INT8** engine ran (23.1 - 15.2) / 15.2 = **~52% faster** than the FP16 one.

**Comment by zldrobit** [comment]
I tested INT8 TensorRT optimization on yolov3-416 models on RTX 2080 Ti with TensorRT 7.0 (batchsize is 64):

| model | FP16 | INT8 |
|--------|-------|------|
| yolov3-416 | 215.9ms | 168.4ms |

The INT8 model is around (215.9-168.4)/215.9 = **22.0% faster** than FP16 one.

**Comment by ttyio** [comment]
Hello @le8888e , I checked your log, are you using maxwell devices? there is no TensorCore so the speed up is limited, and there would be reformat cost when int8 enabled but not full network can run in int8, so we did not get decent perf gain.

Thanks for your perf data sharing @jkjung-avt @zldrobit

**Comment by le8888e** [comment]
Hi @ttyio,

Thank you, but I'm using Tesla V100.

**Comment by ttyio** [comment]
Sorry @le8888e , we only have fp16 tensorcore in V100, and int8 tensorcore only available for Xavier, Turing and Ampere. 
We use IDP, IDP4A on v100, https://docs.nvidia.com/cuda/cuda-binary-utilities/index.html#volta
it not always helps for all cases.

**Comment by le8888e** [comment]
Hi @ttyio ,

What is the proper way to do pre-process images for calibration? Should they be normalized? In my experiments, the output of quantized model differs a lot with different image pre-process methods.

Thank you.

**Comment by ttyio** [comment]
Hello @le8888e ,

Calibration tried to extract data distribution from the calibration dataset, different pre-process would result different intermediate activation data distribution. So we should apply the same pre-process steps as how we handle real input in deployment environment. 

FYI. We have an example how we do PTQ for BERT on SQuAD1.1 dataset
https://github.com/NVIDIA/TensorRT/tree/release/7.2/demo/BERT#evaluating-ptq-post-training-quantization-int8-accuracy-using-the-squad-dataset

**Comment by ttyio** [comment]
close since no activity for more than 3 weeks, please reopen if you still have issue, thanks!

**Comment by wugoukanle** [comment]
you can use trex to draw engine layer with layer precision
